client/SMCE_Client.cpp

```diff
@@ -0,0 +1,339 @@
+#include <SMCE/BoardConf.hpp>                                                                                                    
+#include <SMCE/BoardView.hpp>                                                                                                    
+#include <SMCE/Sketch.hpp>                                                                                                       
+#include <SMCE/SketchConf.hpp>                                                                                                   
+#include <SMCE/Toolchain.hpp>
+                                                                                               
+                                                                                                                                 
+using namespace std::literals;                                                                                                   
+                                                                                                                                                                                                                                                                                                    
+void print_help(const char* argv0){                                                                                              
+    std::cout << "Usage: " << argv0 << " <fully-qualified-board-name> <path-to-sketch>" << std::endl;                            
+}                                                                                                                                
+                                                                                                                                 
+void exit_sketch(int code){                                                                                                      
+    std::cerr << "Error code:  "<< code << std::endl;                                                                            
+}                                                                                                                                
+
+// atomic bool for handling threads
+static std::atomic_bool run_threads = true;
+static std::atomic_bool mute_uart = false;
+static std::mutex t_lock;
+//Listen to the uart input from board, writes what it read to terminal                                                           
+void uart_listener(smce::VirtualUart uart,bool file_write,std::string path){                                                                                      
+    auto tx = uart.tx();
+    std::string file = path+"/uartlog"+get_time()+".txt";                                                                                                         
+    while(run_threads){                                                                                                          
```

Comment:
Look into jthread, it auto joins and uses stop tokens:
https://en.cppreference.com/w/cpp/thread/jthread

using a global ruins the re-usability of this function, in the same vain, t_lock seems like a hack.

So how to resolve this? I would extract the important parts from this, which is collecting the uart buffer and possibly put it in its own class which then has methods like stop() and start(VirtualUart), and then, have it take a callback which calls your method everytime there is new content.